### PR TITLE
Add Enhanced Targeting Features.

### PR DESCRIPTION
**Pull request by ashleysommer**
- Focus Node Filtering.
  - You can now pass in a list of focus nodes to the validator, and it will only validate those focus nodes.
  - Note, you still need to pass in a SHACL Shapes Graph, and the shapes still need to target the focus nodes.
  - This feature will filter the Shapes' targeted focus nodes to include only those that are in the list of specified focus nodes.
  
Update: See comment further down for big additional update to this PR.

### CODE REVIEW

**Comment by ashleysommer** [comment]
Note, I'll leave this PR open for comment for a couple of days, and I want to add the second half of the "targeting" feature-branch to this PR too.
That next feature is the ability to select a specific Shape (Node shape or Property shape) to use, instead of selecting all shapes from the Shapes file.

Then the combination of Focus Node select mode, and Shape select mode will allow powerful targeting that is otherwise not available in PySHACL.

File: pyshacl/validate.py [hunk 1]
```diff
@@ -406,6 +429,7 @@ def validate(
     allow_warnings: Optional[bool] = False,
     max_validation_depth: Optional[int] = None,
     sparql_mode: Optional[bool] = False,
+    focus_nodes: Optional[List[Union[str | URIRef]]] = None,
```
Comment (ajnelson-nist) [inline pyshacl/validate.py:432]: `Union[str | URIRef]` raises a syntax error pre-Python-3.10.  How about `Union[str, URIRef]`?

Comment (ashleysommer) [reply pyshacl/validate.py:432]: Thank you!
I blame that on my new LLM-based code-completion assistant, it must've added the typing in that manner, I don't remember doing so.

**Comment by ashleysommer** [comment]
The rest of the Enhanced Targeting features are landed in this PR.

## Focus Node Filtering, and Shape Selection
- Focus Node Filtering
  - You can pass in a list of focus nodes to the validator, and it will only validate those focus nodes.
  - _Note_, you still need to use a SHACL ShapesGraph, and the Shapes _still need to target_ the focus nodes.
  - This feature will filter the Shapes' targeted focus nodes to include only those that are in the list of specified focus nodes.
- SHACL Shape selection
  - You can pass in a list of SHACL Shapes to the validator, and it will use only those Shapes for validation.
  - This is useful for testing new shapes in your shapes graph, or for many other procedure-driven use cases.
- Combined Shape Selection with Focus Node selection.
  - The combination of the above two new features is especially powerful.
  - If you give the validator a list of Shapes to use, and a list of focus nodes, the validator will operate in
    a highly-targeted mode, it feeds those focus nodes directly into those given Shapes for validation.
  - In this mode, the selected SHACL Shape does not need to specify any focus-targeting mechanisms of its own.

**Comment by nicholascar** [comment]
Looks good @ashleysommer. 

Perhaps we need to see the test suite results in GitHub though? I trust you have them all passing locally but as a PR reviewer, I would like to see them here too, rather than having to run them myself locally too to test.

**Review by nicholascar — APPROVED** [review]

**Comment by ashleysommer** [comment]
> Perhaps we need to see the test suite results in GitHub though?

Yeah, I need to make a change in the PySHACL CI configuration, the Drone VM I used to have set up is no longer running. I'm not a fan of github-actions, but it seems like its the easiest way forward.
And @ajnelson-nist recently created a PR to enable type-checking in Continuous Integration, so that needs to be absorbed in too. https://github.com/RDFLib/pySHACL/pull/250

**Comment by ashleysommer** [comment]
Got whole test suite including linting, type-checking, and tox test matrix working in Github-actions, and merged the upstream changes into this PR, so you will now see the tests run against this PR.

**Comment by ashleysommer** [comment]
I'm merging this now, but I screenshotted the green ticks for reference:
image

**Pull request merged**